### DISCUSSION TITLE
Quarkus integration

### DISCUSSION THREAD

**Discussion by smil2k** [jdbi/jdbi#2618] [Show and tell]
Dear JDBI Community!

I was missing JDBI integration with Quarkus. 

The following repository is in quarkiverse, the home of 3rd party quarkus code:

https://github.com/quarkiverse/quarkus-jdbi

Native compilation already works. I would like to continue to support a similar Spring5 JdbiRepository annotation also. (It would be good to have one common though.)

**Comment by stevenschlansker** [comment]
Hi @smil2k , I do not think any Jdbi developers have experience with Quarkus. What features do you find missing? Are you interested in contributing those features to the project?

**Comment by smil2k** [comment]
Hi @stevenschlansker,

Thank you for your prompt answer!

Quarkus is very similar to spring boot, the notable differences are: 
- Graal VM Native compilation is a primary goal 
- it is based on J2EE standards (microprofile)

JDBI is using reflection extensively, therefore I needed to create a plugin for the quarkus build system to auto-detect those classes using annotation scanning. Quarkus has its own infrastructure for hosting and keeping build plugins up to date. I think this is the proper place for it.

I think the following would be helpful:
- include it in the documentation somehow along with other integrations.
- move the  @JdbiRepository annotation into another package, so the quarkus plugin could also make use of it.

**Reply by hgschmie** [reply]
would you like to provide a PR or should we start to pick up code from there?

While we have been very lenient in the past with taking outside code (e.g. we don't have a formal contributor agreement), pulling in a larger chunk of code without either having some signoff makes me uncomfortable.

**Reply by smil2k** [reply]
I think quarkivers is a good location for building and release, as it runs all the builds and generates all the documentation to the quarkus documentation. This is how other projects are connected. 

The quarkus adaptation is not large at all. Please look at the github project, it is about 1000 lines including imports and the rest.

What I mentioned about JdbiRepository annotation, is in the spring-boot support, to mark the auto generated DAOs. I think it would make sense to make some common project which adds runtime dependencies to the project at hand like the annotation or probably some Jdbi instance customization like this: https://lankydan.dev/providing-your-own-jackson-objectmapper-in-quarkus. So these things could be portable between spring boot and quarkus.

**Reply by stevenschlansker** [reply]
We are already working on better Graal Native support over in https://github.com/jdbi/jdbi/pull/2476, so that might be part of the solution here.

**Reply by stevenschlansker** [reply]
Once we complete that support, Jdbi should include the necessary metadata for its own classes to run Graal Native without annotation or detection. But there's no specific timeframe for integration yet as we have found some bugs in the `native-image-maven-plugin`.

**Reply by smil2k** [reply]
That is fantastic. 

I think the hardest part was for me to create examples for all usages. Native compilation is very tricky, so a maintained "kitchen sink application" for native compilation would be very good.

**Reply by stevenschlansker** [reply]
Yes, I definitely appreciate that it's hard to get started with Jdbi + native-image. It's really early days for all of this, and there are lots of bugs and kinks to work out! I have personally spent dozens of hours to get it working and am not done yet. So any additional help you or anyone else can offer to make the experience better is much appreciated.
